contracts/plugins/assets/erc20/RewardableERC20.sol

```diff
@@ -36,39 +43,59 @@ abstract contract RewardableERC20 is IRewardable, ERC20 {
         _claimAccountRewards(msg.sender);
     }
 
+    function sync() external {
+        _claimAndSyncRewards();
+    }
+
     function _syncAccount(address account) internal {
         if (account == address(0)) return;
-        uint256 shares = balanceOf(account);
         uint256 accountRewardsPerShare = lastRewardsPerShare[account];
-        if (rewardsPerShare == accountRewardsPerShare) return;
-        uint256 delta = rewardsPerShare - accountRewardsPerShare;
-
-        // {qRewards} = {qRewards/share} * {qShare} / {qShare/share}
-        uint256 newRewards = (delta * shares) / one;
-        lastRewardsPerShare[account] = rewardsPerShare;
-        accumulatedRewards[account] += newRewards;
+        uint256 shares = balanceOf(account);
+        uint256 _accumuatedRewards = accumulatedRewards[account];
+        uint256 _rewardsPerShare = rewardsPerShare;
+        if (accountRewardsPerShare < _rewardsPerShare) {
+            uint256 delta = _rewardsPerShare - accountRewardsPerShare;
+            delta = (delta * shares) / one;
+            // {qRewards} = {qRewards/share} * {qShare} / {qShare/share}
+            _accumuatedRewards = _accumuatedRewards + delta;
+        }
+        lastRewardsPerShare[account] = _rewardsPerShare;
+        accumulatedRewards[account] = _accumuatedRewards;
     }
 
-    function _claimAndSyncRewards() internal {
-        uint256 delta;
+    function _claimAndSyncRewards() internal virtual {
         uint256 _totalSupply = totalSupply();
-        if (_totalSupply > 0) {
-            uint256 initialBal = rewardToken.balanceOf(address(this));
-            _claimAssetRewards();
-            uint256 endingBal = rewardToken.balanceOf(address(this));
-            delta = endingBal - initialBal;
+        if (_totalSupply == 0) {
+            return;
+        }
+        _claimAssetRewards();
+        uint256 balanceAfterClaimingRewards = rewardToken.balanceOf(address(this));
+
+        uint256 _rewardsPerShare = rewardsPerShare;
+        uint256 _previousBalance = previousBalance;
 
+        if (balanceAfterClaimingRewards > _previousBalance) {
+            uint256 delta = balanceAfterClaimingRewards - _previousBalance;
             // {qRewards/share} += {qRewards} * {qShare/share} / {qShare}
-            rewardsPerShare += ((delta) * one) / _totalSupply;
+            _rewardsPerShare += (delta * one) / _totalSupply;
         }
+        previousBalance = balanceAfterClaimingRewards;
+        rewardsPerShare = _rewardsPerShare;
     }
 
     function _claimAccountRewards(address account) internal {
         uint256 claimableRewards = accumulatedRewards[account] - claimedRewards[account];
+        if (claimableRewards == 0) {
+            return;
+        }
+
         emit RewardsClaimed(IERC20(address(rewardToken)), claimableRewards);
-        if (claimableRewards == 0) return;
```

Comment:
i guess this is up for debate but i think we always want to emit `RewardsClaimed` even if the amount is 0. @tbrent thoughts?

Reply:
I moved up above the if statement.